### PR TITLE
create an auto strategy that can attempt to work out domain and expos…

### LINKED ISSUE #5: have a startup mode of 'auto'?
to make a single YAML that can startup fabric8 on kubernetes, the main 2 parts of parameterization required are `domain` if folks use ingress/router and `expose-rule`.

I wonder if we could default expose-rule to `auto` which would then reuse the same logic in gofabric8 to default the mode? (e.g. try detect mini* and use nodePort etc otherwise if it spots a router / ingress-controller use that otherwise default to load-balancer).

Am just pondering how we can make fabric8 as easy to install as possible on vanilla kubernetes

### PR DESCRIPTION
**Pull request by rawlingsj**
…er configs fixes #5

### CODE REVIEW

**Comment by rawlingsj** [comment]
@jstrachan @jimmidyson pr ready with the comments from Jimmi including to use an auto strategy which feels really nice now.  

Over time we can add support for more routers or default configs given the platform exposecontroller is running on but this is a start.

I wont do anything to this PR until I can test it on AWS which should be later using the single yaml file thats being worked on.

File: exposestrategy/ingress.go [hunk 1]
```diff
@@ -33,7 +33,12 @@ func NewIngressStrategy(client *client.Client, encoder runtime.Encoder, domain s
 	}
 
 	if len(domain) == 0 {
-		return nil, errors.New("domain is required")
+		domain, err = getAutoDefaultDomain(client)
+		if err != nil {
+			errors.Wrap(err, "failed to get a domain")
```
Comment (jimmidyson) [inline exposestrategy/ingress.go]: should this be return?

File: exposestrategy/ingress.go [hunk 2]
```diff
@@ -33,7 +33,12 @@ func NewIngressStrategy(client *client.Client, encoder runtime.Encoder, domain s
...
+		} else {
```
Comment (jimmidyson) [inline exposestrategy/ingress.go]: drop the else once you've added the return back in above

File: exposestrategy/route.go [hunk 3]
```diff
@@ -36,7 +37,12 @@ func NewRouteStrategy(client *client.Client, oclient *oclient.Client, encoder ru
 	}
 
 	if len(domain) == 0 {
-		return nil, errors.New("domain is required")
+		domain, err = getAutoDefaultDomain(client)
+		if err != nil {
+			errors.Wrap(err, "failed to get a domain")
```
Comment (jimmidyson) [inline exposestrategy/route.go]: return

File: exposestrategy/route.go [hunk 4]
```diff
@@ -36,7 +37,12 @@ func NewRouteStrategy(client *client.Client, oclient *oclient.Client, encoder ru
...
+		} else {
```
Comment (jimmidyson) [inline exposestrategy/route.go]: drop else

File: exposestrategy/strategy.go [hunk 5]
```diff
@@ -58,7 +58,13 @@ func New(exposer, domain string, client *client.Client, restClientConfig *restcl
 			return nil, errors.Wrap(err, "failed to create ingress expose strategy")
 		}
 		return strategy, nil
+	case "":
```
Comment (jimmidyson) [inline exposestrategy/strategy.go]: what about allowing "auto" as well for explicitly enabling auto?

File: exposestrategy/auto.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	nodePort     = "nodeport"
+	route        = "route"
+	domainExt    = ".xip.io"
+)
+
+func NewAutoStrategy(exposer, domain string, client *client.Client, restClientConfig *restclient.Config, encoder runtime.Encoder) (ExposeStrategy, error) {
+
+	exposer, err := getAutoDefaultExposeRule(client)
+	if err != nil {
+		errors.Wrap(err, "failed to automatically get exposer rule.  consider setting 'exposer' type in config.yml")
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: return

File: exposestrategy/auto.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	} else {
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: drop else

File: exposestrategy/auto.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+		glog.Infof("Using exposer strategy: %s", exposer)
+	}
+
+	// only try to get domain if we need wildcard dns and one wasn't given to us
+	if len(domain) == 0 && (strings.EqualFold(ingress, exposer) || strings.EqualFold(route, exposer)) {
+
+		domain, err = getAutoDefaultDomain(client)
+		if err != nil {
+			errors.Wrap(err, "failed to get a domain")
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: return

File: exposestrategy/auto.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+		} else {
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: drop reuse

File: exposestrategy/strategy.go (hunk 5 above)
Comment (rawlingsj) [inline exposestrategy/strategy.go]: we decided on IRC to drop that

File: exposestrategy/auto.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	return New(exposer, domain, client, restClientConfig, encoder)
+}
+
+func getAutoDefaultExposeRule(c *client.Client) (string, error) {
+
+	nodes, err := c.Nodes().List(api.ListOptions{})
+	if err != nil {
+		errors.Wrap(err, "failed to find any nodes")
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: return

File: exposestrategy/auto.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	if len(nodes.Items) == 1 {
+		node := nodes.Items[0]
+		if node.Name == "minishift" || node.Name == "minikube" {
+			return nodePort, nil
+		}
+	}
+
+	t, err := typeOfMaster(c)
+	if err != nil {
+		return "", err
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: wrap error to give context

File: exposestrategy/auto.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+		errors.Wrap(err, "failed to find any nodes")
+	}
+
+	// if we're mini* then there's only one node, any router / ingress controller deployed has to be on this one
+	if len(nodes.Items) == 1 {
+		node := nodes.Items[0]
+		if node.Name == "minishift" || node.Name == "minikube" {
+			ip, err := getExternalIP(node)
+			if err != nil {
+
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: remove blank line

File: exposestrategy/auto.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+				return "", err
+			}
+			return ip + domainExt, nil
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: what about if domain ext doesn't begin with .?

File: exposestrategy/auto.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+	// check for a gofabric8 ingress labelled node
+	selector, err := unversioned.LabelSelectorAsSelector(&unversioned.LabelSelector{MatchLabels: map[string]string{"fabric8.io/externalIP": "fabric8"}})
+	nodes, err = c.Nodes().List(api.ListOptions{LabelSelector: selector})
+	if len(nodes.Items) == 1 {
+		node := nodes.Items[0]
+		ip, err := getExternalIP(node)
+		if err != nil {
+			return "", err
+		}
+		return ip + domainExt, nil
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: as above

File: exposestrategy/auto.go [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+	// look for a stackpoint HA proxy
+	pod, _ := c.Pods("stackpoint-system").Get("spc-balancer")
+	if pod != nil {
+		containers := pod.Spec.Containers
+		for _, container := range containers {
+			if container.Name == "spc-balancer" {
+				for _, e := range container.Env {
+					if e.Name == "BALANCER_IP" {
+						return e.Value + domainExt, nil
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: as above

File: exposestrategy/auto.go [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+		node := nodes.Items[0]
+		ip, err := getExternalIP(node)
+		if err != nil {
+			return "", err
+		}
+		return ip + domainExt, nil
+	}
+
+	// look for a stackpoint HA proxy
+	pod, _ := c.Pods("stackpoint-system").Get("spc-balancer")
```
Comment (jimmidyson) [inline exposestrategy/auto.go]: good to use consts for these strings

File: exposestrategy/auto.go (hunk 13 above)
Comment (rawlingsj) [inline exposestrategy/auto.go]: domainExt is a const = ".xip.io", if this is needs to be overridden then user can add a domain value to the config.  readme has been updated to explain the default of xip

File: exposestrategy/auto.go (hunk 6 above)
Comment (rawlingsj) [inline exposestrategy/auto.go]: thanks done

File: exposestrategy/route.go (hunk 4 above)
Comment (rawlingsj) [inline exposestrategy/route.go]: thanks done

File: exposestrategy/ingress.go (hunk 2 above)
Comment (rawlingsj) [inline exposestrategy/ingress.go]: thanks done

File: exposestrategy/ingress.go (hunk 1 above)
Comment (rawlingsj) [inline exposestrategy/ingress.go]: thanks done

**Comment by rawlingsj** [comment]
@jimmidyson think that's everything, let me know if not as I try and get a fabric8-devops release out this evening

File: controller/config.go [hunk 17]
```diff
@@ -4,15 +4,17 @@ import (
...
+	"github.com/golang/glog"
 	"github.com/pkg/errors"
 	"gopkg.in/v2/yaml"
 )
 
 func LoadFile(path string) (*Config, error) {
 	content, err := ioutil.ReadFile(path)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to read config file")
+		glog.Infof("No %s file found.  Will try to figure out defaults", path)
```
Comment (jimmidyson) [inline controller/config.go]: if we want config file to be optional you'll need to check if previous error is `err != nil && !os.IsNotExist(err)`

File: controller/config.go [hunk 18]
```diff
@@ -4,15 +4,17 @@ import (
...
 	}
+
 	c, err := Load(string(content))
```
Comment (jimmidyson) [inline controller/config.go]: load requires some content so requires a file to read from above

**Comment by jimmidyson** [comment]
Sorry for late comments. That should be it now.

File: controller/config.go (hunk 18 above)
Comment (jimmidyson) [inline controller/config.go]: update load to create a default config (auto) & decode into that default object

File: controller/config.go (hunk 17 above)
Comment (rawlingsj) [inline controller/config.go]: done

File: controller/config.go (hunk 18 above)
Comment (rawlingsj) [inline controller/config.go]: are you sure?  I've been testing this and it seems to work fine

Comment (rawlingsj) [inline controller/config.go]: sorry not completely following this one, can you elaborate?  we don't use auto anymore as we try to work out the automated defaults if domain or exposer are empty.  I've tested this and it seems to be working fine with 
- empty domain + exposer 
- no config.xml 
- populated config.xml using domain and exposer overrides

all seems to be working great

**Comment by rawlingsj** [comment]
No worries about the late comments - all good and thanks for them! 

I didn't quite get the last comment but as I said it's all tested.  I'm totally happy to pick up anything I've missed tomorrow but I'm going to merge this so that we can kick off a full release which is needed tonight.  

I dont think any of the outstanding items cause risk after my testing but like I said we can address that in another release tomorrow (or soon) if need be.

File: controller/config.go (hunk 18 above)
Comment (jimmidyson) [inline controller/config.go]: ah looks like i had already handled the default config & empty file content... go me. looks fine then.

**Comment by rawlingsj** [comment]
yay thanks for the feedback @jimmidyson!

**Pull request merged**